python/ray/_private/runtime_env/pip.py

```diff
@@ -428,16 +433,25 @@ async def create(
         target_dir = self._get_path_from_hash(hash)
 
         async def _create_for_hash():
-            await PipProcessor(target_dir, runtime_env, logger)
+            await PipProcessor(
+                target_dir,
+                runtime_env,
+                logger,
+            )
 
             loop = get_running_loop()
             return await loop.run_in_executor(
                 None, get_directory_size_bytes, target_dir
             )
 
-        self._creating_task[hash] = task = create_task(_create_for_hash())
-        task.add_done_callback(lambda _: self._creating_task.pop(hash, None))
-        return await task
+        if uri not in self._create_locks:
+            # async lock to prevent the same virtualenv being concurrently installed
+            self._create_locks[uri] = asyncio.Lock()
+
+        async with self._create_locks[uri]:
+            self._creating_task[hash] = task = create_task(_create_for_hash())
+            task.add_done_callback(lambda _: self._creating_task.pop(hash, None))
+            return await task
```

Comment:
why do we need this callback? the lock should get released when `await task` is finished

Reply:
Ah this is existing code, I just wrapped it in the lock.  The only place the `_creating_task` dict is used is in `delete_uri`, where current running creation tasks are canceled if `delete_uri` is called.  I think it makes sense to leave it in.